### PR TITLE
Sharing and interactive visualization chapter

### PR DESCRIPTION
**Pull request by kevinrue**
A first attempt to kick off things.

### CODE REVIEW

**Comment by LTLA** [comment]
Formatting comments:

- What's up with the triple dashes? Single dashes aren't good enough for you?
- Keep an eye on the complexity of your sentences. Some of them could be broken up.
- Try not to use package names as section headings, instead describe what it does for people, e.g., "Exploring SummarizedExperiment objects".
- I'd prefer to not stick `images/iSEE-dataflow.png` in the repo. But I'll grudgingly accept it. Note that you'll have to modify `rebuild.sh` in the `Orchestrating` repo to `cp` the `images` dir.

Also, how are you planning to show the interactiveness? We could put `iframes` in the book pointing to a shiny server somewhere, if only we had a Shiny server that we could use. Probably better than having static screenshots.

**Comment by kevinrue** [comment]
* I know double and triple dashes turn into en- and em dash, though I can never remember which ones I'm supposed to use. Probably best to stick with single dash in that case.
* I'm generally aware of the complexity of sentences. That said, a helpful advice I got from my PhD supervisors is that to get started, anything is good. Cleaning up is easier when something's there in the first place.
* Good point.
* In my defense, the image was there when I picked up the chapter. Not a problem for me to remove it. I may even have left it influence me into spending/wasting time thinking what text I could write to lead to the figure.

That is the question isn't it? Part of my issue writing anything comes exactly from that point of "how are we illustrating it". We'll think of something and then I'm sure this chapter will write itself.

**Comment by LTLA** [comment]
> Cleaning up is easier when something's there in the first place.

What?! You mean it's not perfect to start with? 🇳🇿 👎 

> We'll think of something and then I'm sure this chapter will write itself.

I reckon we should go with `iframe`s, plus the commands required to generate them. Anything else would be crappy. Now we just have to find a sucker with a Shiny server who can host this for us.

File: analysis/interactive.Rmd [hunk 1]
```diff
@@ -17,12 +17,82 @@ source("workflows/extractor.R")
...
+if (interactive()) {
+    iSEE(sce.pbmc)
+}
+```
+
+For examples of the iSEE interface applied to various types of genomics datasets, see https://github.com/csoneson/iSEE, https://github.com/LTLA/iSEE2018, https://github.com/federicomarini/iSEE_instances.
+
+To get started, iSEE apps each provide a guided tour to walk the user through the functionality of the interface.
+
+For analysts generating the `sce` object, we recommend saving the `sce` via one of the following methods :
```
Comment (LTLA) [inline analysis/interactive.Rmd]: This part about data in/out seems a bit irrelevant. Better suited to the `import-export.Rmd` section.

File: analysis/interactive.Rmd [hunk 2]
```diff
@@ -4,90 +4,259 @@ output:
...
+We note that these are examples only; we encourage readers to consider them as templates for learning and developing applications adapted to their own specific needs and purposes.
+
+### Quality control
+
+Having previously computed quality control metrics (see section [Quick start](#interactive-quickstart) above), it is possible to launch an app instance that immediately displays a set of panels preconfigured to focus on selected quality control metrics.
+
+For example, a common view for initial inspection of single-cell RNA-sequencing data sets generated by microfluidics platforms plots the library size of each cell, in decreasing order.
+An elbow in this plot generally reveals the transition between good quality cells and low quality cells or empty droplets.
+
+Another common view is to overlay the library size (in log space) as a color gradient over the result of a dimensionality reduction method (e.g., t-SNE, UMAP, ivis).
```
Comment (federicomarini) [inline analysis/interactive.Rmd]: Do we want to add refs to the three dim red methods?

File: analysis/interactive.Rmd [hunk 3]
```diff
@@ -4,90 +4,259 @@ output:
...
+These are unlikely to capture the full information content of a complex scRNA-seq data set, and do not allow further exploration by the reader.
+To complement this traditional mode of dissemination, it is becoming increasingly common to deploy web-based interactive data browsers, allowing interested readers to investigate their respective hypotheses using the published data.
 
-## Advanced Usage
+With `iSEE`, such interactive exploration can be guided by small guided tours - step-by-step walkthroughs of the different panels with pointers to facilitate their interpretation, from the authors' point of view.
+At any time, the viewer is still free to leave the provided tour and explore the data from their own perspective.
+All that is needed to add a tour to an `iSEE` instance is a data frame with two columns; the first column containing the UI element to be pointed to in each step of the tour, and the second one containing the text to display at that step.
+This data frame must then be provided to the `iSEE()` function via the `tour` argument. Examples of tours can be found in https://github.com/ltla/iSEE2018.
```
Comment (federicomarini) [inline analysis/interactive.Rmd]: Should be `LTLA` all caps - although it does not make a difference in term of URL

File: analysis/interactive.Rmd [hunk 4]
```diff
@@ -4,90 +4,259 @@ output:
...
+Practically, investigating the expression levels of known marker genes is a common approach to assigning cell type labels to inferred clusters.
+
+### Collaborative analysis
+
+Analysis and interpretation of complex data sets typically involves multiple analysts.
+Rather than sharing analysis results via multiple versions of static figures and reports, it is often helpful to share an interactive version of a plot, allowing the other involved parties to independently explore the data in further depth.
+
+With `iSEE`, this can be achieved at any time by clicking on the "Display panel settings" button under the "Diagnostics" tab in the top right corner, copying the displayed code into an R script, and sharing this chunk of code with collaborators.
+However, note that they will also need to have access to a copy of the original `sce` object used for the analysis.
+In that setup, executing the code in the script and launching iSEE with the following command will open an instance mimicking what was shown on the screen when the "Display panel settings" button was clicked.
```
Comment (federicomarini) [inline analysis/interactive.Rmd]: `iSEE` to replace iSEE

File: analysis/interactive.Rmd [hunk 5]
```diff
@@ -4,90 +4,259 @@ output:
...
+For instance, users may select cells in one panel and choose to highlight the selected cells in the other using the "Selection parameters" collapsible box, in an effort to determine an adequate threshold on the library size to filter cells to retain for downstream analyses.
+
+Users may also change the information displayed in any panel. For instance, users may choose to color data points by percentage of UMI mapped to mitochondrial genes (“pct_counts_Mito”) in the “Reduced dimension plot 1”.
+Using the transfer of point selection between panels, users could select cells with small library sizes (in the “Column data plot 1” panel) and highlight them in the “Reduced dimension plot 1” panel, to investigate a possible relation between library size, clustering, and proportion of reads mapped to mitochondrial genes.
+
+### Annotation and identification of cell populations
+
+Recent efforts such as the Human Cell Atlas (HCA) and the NIH Human BioMolecular Atlas Program (HuBMAP) have initiated the creation of reference maps of cell types and states in the human body, in health and disease.
```
Comment (federicomarini) [inline analysis/interactive.Rmd]: We could/should add the websites for the HCA and HuBMAP?

File: analysis/interactive.Rmd (hunk 2 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: Done

**Review by federicomarini — APPROVED** [review]

File: analysis/interactive.Rmd (hunk 3 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: Done (d1814524)

File: analysis/interactive.Rmd (hunk 4 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: I caught this one and others leftover ones (d1814524)

File: analysis/interactive.Rmd (hunk 5 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: Done (d1814524)

File: analysis/interactive.Rmd [hunk 6]
```diff
@@ -4,21 +4,265 @@ output:
...
+Rather than sharing analysis results via multiple versions of static figures and reports, it is often helpful to share an interactive version of a plot, allowing the other involved parties to independently explore the data in further depth.
+
+With `iSEE`, this can be achieved at any time by clicking on the "Display panel settings" button under the "Diagnostics" tab in the top right corner, copying the displayed code into an R script, and sharing this chunk of code with collaborators.
+However, note that they will also need to have access to a copy of the original `sce` object used for the analysis.
+In that setup, executing the code in the script and launching `iSEE` with the following command will open an instance mimicking what was shown on the screen when the "Display panel settings" button was clicked.
+It is important to note that all `iSEE` applications remain interactive irrespective of their initial configuration; only the initial layout and state of the application is modified from the default setup. 
+
+```{r, eval=FALSE}
+iSEE(sce, redDimArgs = redDimPlotArgs, colDataArgs = colDataPlotArgs,
```
Comment (LTLA) [inline analysis/interactive.Rmd]: This is pretty confusing if `redDimPlotArgs` and friends are never actually created.

File: analysis/interactive.Rmd [hunk 7]
```diff
@@ -4,21 +4,265 @@ output:
...
+
+Another common view is to overlay the library size (in log space) as a color gradient over the result of a dimensionality reduction method such as t-SNE, UMAP, or ivis [@vanDerMaaten2008tsne;@mcInnes2018umap;@szubert2019ivis].
+This view may reveal trajectories or clusters associated with library size.
+Depending on the experimental context, such observations may result from technical or biological phenomena.
+For instance, it may indicate that normalization was not completely effective at removing cell biases.
+Alternatively, it could also indicate the presence of multiple cell types that differ in total RNA content.
+
+In the example below, we demonstrate that an `iSEE` app can be preconfigured to immediately display the views described above.
+
+```{r, eval=FALSE}
```
Comment (LTLA) [inline analysis/interactive.Rmd]: I suggest we isolate the set up from the `runApp()`, and only have the latter in an `eval=FALSE` block. This ensures that we verify that the majority of the set-up code is runnable.

File: analysis/interactive.Rmd [hunk 8]
```diff
@@ -4,21 +4,265 @@ output:
...
+```{r}
+library(scran)
+
+# Compute cell clusters
+g <- buildSNNGraph(sce.pbmc, k=10, use.dimred = 'PCA')
+clust <- igraph::cluster_walktrap(g)$membership
+sce.pbmc$cluster <- factor(clust)
+
+# Identify positive markers for each cluster
+markers.pbmc.up <- findMarkers(sce.pbmc, sce.pbmc$cluster, direction="up", log.p=TRUE)
```
Comment (LTLA) [inline analysis/interactive.Rmd]: This now has a `sorted=FALSE` option.

File: analysis/interactive.Rmd (hunk 6 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: I've removed the code chunk. Resolving this.

Comment (kevinrue) [reply analysis/interactive.Rmd]: Code chunk removed in 147c21c

File: analysis/interactive.Rmd (hunk 7 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: I've split the relevant chunks in 147c21c

File: analysis/interactive.Rmd (hunk 8 above)
Comment (kevinrue) [reply analysis/interactive.Rmd]: Code chunk updated in 147c21c

File: ref.bib [hunk 9]
```diff
@@ -1058,3 +1058,50 @@ @Article{butler2018integrating
    Month="06"
 }
 
+@article{vanDerMaaten2008tsne,
```
Comment (LTLA) [inline ref.bib]: This is redundant, we have one of these already surely.

File: analysis/interactive.Rmd [hunk 10]
```diff
@@ -4,21 +4,256 @@ output:
...
+## Reproducibility
+
+Although this chapter focuses on interactive exploration of data, we note that it is important to retain reproducibility, and to have a record of how each figure was generated.
+With `iSEE`, this information is readily available via the "Export R code" button under the "Diagnostics" tab in the top-right corner of the GUI.
+At any time, copying the code displayed in the modal window and executing it in the R session from which the `iSEE` app was launched exactly reproduces all plots currently displayed in the GUI.
+
+## Additional resources
+
+- Publication (F1000Research, 2018): https://f1000research.com/articles/7-741/v1
+- Source code: https://github.com/csoneson/iSEE
```
Comment (LTLA) [inline analysis/interactive.Rmd]: We don't need these two things above. Just talk about the deployed examples.

Comment (kevinrue) [reply analysis/interactive.Rmd]: Done in e1f154de

Comment (LTLA) [reply analysis/interactive.Rmd]: You might want to expand a bit more on what you mean by deployed instances. Just have a few sentences outside of a list, and then just list out the ones we have.
